src/runtime_src/core/edge/drm/zocl/sched_exec.c

```diff
@@ -2784,6 +2784,53 @@ static irqreturn_t sched_cq_isr(int irq, void *arg)
 	return IRQ_HANDLED;
 }
 
+/**
+ * zocl_exec_reset() 
+ * 
+ * This function ensures that the device exec_core state is reset to
+ * same state as was when scheduler was originally probed for the device.
+ * This gets called after each xclbin load. This gets called for the first time also
+ *
+ * @drm: Device node to initialize
+ *
+ */
+int
+zocl_exec_reset(struct drm_device *drm)
+{
+	unsigned int i;
+	struct drm_zocl_dev *zdev = drm->dev_private;
+	struct sched_exec_core *exec_core = zdev->exec;
+
+	exec_core->scheduler = &g_sched0;
+	exec_core->num_slots = 16;
+	exec_core->num_cus = 0;
+	exec_core->cu_base_addr = 0;
+	exec_core->cu_shift_offset = 0;
+	exec_core->polling_mode = 1;
+	exec_core->cq_interrupt = 0;
+	exec_core->configured = 0;
+	exec_core->cu_isr = 0;
+	exec_core->cu_dma = 0;
+	exec_core->num_slot_masks = 1;
+	exec_core->num_cu_masks = 0;
+	exec_core->ops = &penguin_ops;
+
+	for (i = 0; i < MAX_SLOTS; ++i)
+		exec_core->submitted_cmds[i] = NULL;
+
+	for (i = 0; i < MAX_U32_SLOT_MASKS; ++i)
+		exec_core->slot_status[i] = 0;
+
+	for (i = 0; i < MAX_U32_CU_MASKS; ++i) {
+		exec_core->cu_status[i] = 0;
+		exec_core->cu_init[i] = 0;
+		exec_core->cu_valid[i] = 0;
+	}
+
```

Comment:
Hi, thanks this is a simpler way to fix one issue that single user use different xclbin which may bring different configuration. So we will have to re-configure the scheduler to make sure the following xclbin won't be broken. What is missing is that, we also want to make sure the previous runs have been finished, if it is hung for long time, user should take care of the situation before loading the next new xclbin, aka cu drainage. Those will be bring by context switch. I have a code in testing mode: 	https://github.com/xdavidz/XRT1/commit/04155fbb72ead68adefaebc2413e7b200db3e65f

But for this effort, I think if we really want to reset the scheduler as of now, please avoid dup code, the following code is due with just the code right below in sched_init_exec. And, most importantly there are some hard-code should be refactored in the near future (like num_slots = 16), those should be from the device-tree or vise versa, but single way. If we have different device-tree config, the driver will just panic today. My advise would be to use a function to abstract common code both in sched_reset and sched_init or even sched_fini. The above context switch code could be an example, again it is being tested. What do you think?
 
exec_core->scheduler = &g_sched0;
	exec_core->num_slots = 16;		exec_core->num_slots = 16;
	exec_core->num_cus = 0;		exec_core->num_cus = 0;
	exec_core->cu_base_addr = 0;		exec_core->cu_base_addr = 0;
	exec_core->cu_shift_offset = 0;		exec_core->cu_shift_offset = 0;
	exec_core->polling_mode = 1;		exec_core->polling_mode = 1;
	exec_core->cq_interrupt = 0;		exec_core->cq_interrupt = 0;
	exec_core->cu_isr = 0;		exec_core->cu_isr = 0;
	exec_core->cu_dma = 0;		exec_core->cu_dma = 0;
	exec_core->num_slot_masks = 1;		exec_core->num_slot_masks = 1;
	exec_core->num_cu_masks = 0;		exec_core->num_cu_masks = 0;
	exec_core->ops = &penguin_ops;		exec_core->ops = &penguin_ops;

Reply:
I see a lot common code in zocl_exec_reset(struct drm_device *drm) and sched_init_exec(struct drm_device *drm).

Could we create a new function like:
inline void init_exec(struct sched_exec_core *exec_core) {
    exec_core->scheduler = &g_sched0;
    exec_core->num_slots = 16;
    ...
}

Reply:
@xdavidz , These changes are good enough for the single process cases. We should reset the scheduler after loading the new xclbin anyway. It is better if we push these changes sothat DSV can validate these cases too.

Reply:
I agree with you that context switch change is the right change to cover all the cases, but as this support may take some more time, i would suggest, we should go with atleast resetting the scheduler changes.

Reply:
thanks, I am ok with ahead with the fix, please make sure no panic during the test. Code looks ok.